### DISCUSSION TITLE
Should Energy Dashboards handle prices in cents

### DISCUSSION THREAD

**Discussion by ZephireNZ** [home-assistant/architecture#857] [General]
Related Issue: https://github.com/home-assistant/core/issues/85414

My integration provides a sensor for power price in units `¢/kWh`. This is the same unit which the company provides on the API, their bills and self-service portals, etc. So it made sense for me to use that unit in the integration.

However, for the `_cost` sensor created provided by the `energy` integration, it incorrectly assumes that all energy price entities are in the currency specified at the server level (eg USD or NZD). This is causing the energy dashboards to show dollar figures 100x larger than they actually are.

https://github.com/home-assistant/core/blob/dev/homeassistant/components/energy/sensor.py#L450

A simple fix that can be done is to change the sensor's unit (either as an code change, or simply a template sensor), but I think it makes more sense that the energy integration should handle cents natively.

I originally assumed there'd be a library that can handle it for us (eg give the currency, the amount with the unit, and it'd give back the price in the base unit) - however it appears to be even simpler than that. Only two non-decimal currencies in the world exist now ([ref](https://en.wikipedia.org/wiki/Non-decimal_currency)), and so the (somewhat obvious) definition of a cent is simply 1/100 of the base currency.

There will be some currencies which have a cent called by a different name (eg instead of using `¢` as unit, a Chinese fēn (分)). I assume some logic could be build, similar to power units, which converts a cent/non-base unit into the base currency unit for any integrations which don't use $/¢ - is there any we are aware of?

Given the above, it seems to me that it's a fair assumption that we can simply divide the sensor value by 100 if the sensor's first unit (ie ¢ in the above) is cents - and future proofing it by making it a conversion function to be able to handle some other cent-like units.

**Comment by parautenbach** [comment]
Just a general remark as someone who has worked on payment and financial systems: It's a common/best practice to always work in and store the minor denomination (cents) and treat it otherwise as a presentation issue. That way, you never have to divide, since division leads to rounding errors. This is under the assumption you don't have fractional cents.

**Reply by AJediIAm** [reply]
Some information which may be useful:

In the Netherlands, energy tarifs and taxes are specified to 5 digits (€ 0.0xxxx).
Dynamic energy prices are provided in €/MWh (€ xx.xx)

The government seems to prefer €/kWh
Most energie providers seem to prefer c/kWh

**Comment by zatricky** [comment]
@ZephireNZ The Nordpool Integration (installable via HACS) is an example of an integration that uses "cents" that aren't cents. It gives the currency part in "öre", the equivalent of cents for SEK / Swedish Crowns - https://github.com/custom-components/nordpool

<img src="https://github.com/user-attachments/assets/aae492dd-821d-4a44-b135-d0d6f3062d4c" width="240">

As a workaround I used a [template](https://github.com/home-assistant/core/issues/85414#issuecomment-1376726496) to convert the price into SEK ; however it would certainly have been convenient to be able to do it without the template.

<img src="https://github.com/user-attachments/assets/4a7be76a-5d59-4ff8-a806-10e0f6b47b94" width="240">

---

Personally, I feel like a generic conversion system would go a long way towards resolving issues like this. I realise it *is* a big ask - but also it would make a huge class of problems just disappear.

As a relevant "pie-in-the-sky" example where I would want to be able to calculate the cost of running an appliance over a month, the calculation would require `SEK/kWh`, `kW`, and `hours` inputs and want an output of SEK. We could have data inputs of öre/kWh, watts, and minutes, and we'd get a final value of SEK, by just plugging these inputs/outputs into the generic conversion function.

On paper, the only data we need to do these calculations ourselves is rather simple:

* 1 SEK = 100 öre
* 1 kW = 1000 W
* 1 hour = 60 minutes
* kWh = kW * hours
* SEK/kWh = SEK / kWh
